HARK/dcegm.py

```diff
@@ -8,428 +8,298 @@
         # Rows are individual crossing points.
-        segments = np.stack([optIdx[idx_change], optIdx[idx_change + 1]], axis=1)
-
-        # Values of both segments on the left point
-        left_v = np.stack(
-            [condVs[idx_change, segments[:, 0]], condVs[idx_change, segments[:, 1]]],
-            axis=1,
-        )
-        # and the right point
-        right_v = np.stack(
-            [
-                condVs[idx_change + 1, segments[:, 0]],
-                condVs[idx_change + 1, segments[:, 1]],
-            ],
-            axis=1,
-        )
+        segments = np.stack((opt_idx[idx_change], opt_idx[idx_change + 1]), axis=1)
+
+        # Get values of segments on both the left and the right
+        left_y = np.zeros_like(segments, dtype=np.float64)
+        right_y = np.zeros_like(segments, dtype=np.float64)
+
+        for i, idx in enumerate(idx_change):
+
+            left_y[i, 0] = cond_ys[segments[i, 0], idx]
+            left_y[i, 1] = cond_ys[segments[i, 1], idx]
+
+            right_y[i, 0] = cond_ys[segments[i, 0], idx + 1]
+            right_y[i, 1] = cond_ys[segments[i, 1], idx + 1]
 
         # A valid crossing must have both switching segments well defined at the
         # encompassing gridpoints. Filter those that do not.
-        valid = np.logical_and(
-            ~np.isnan(left_v).any(axis=1), ~np.isnan(right_v).any(axis=1)
-        )
+        valid = np.repeat(False, len(idx_change))
+        for i in range(len(valid)):
+            valid[i] = np.logical_and(
+                ~np.isnan(left_y[i, :]).any(), ~np.isnan(right_y[i, :]).any()
+            )
 
-        segments = segments[valid, :]
-        switchMs = switchMs[valid, :]
-        left_v = left_v[valid, :]
-        right_v = right_v[valid, :]
+        if not np.any(valid):
 
-        # Find crossing points. Returns a list (m,v) tuples. Keep m's only
-        xing_points = [
-            calc_linear_crossing(switchMs[i, :], left_v[i, :], right_v[i, :])
-            for i in range(segments.shape[0])
-        ]
+            # If there are no valid crossings, return empty arrays.
+            points = np.zeros((0, 2), dtype=np.float64)
+            segments = np.zeros((0, 2), dtype=np.int64)
+            return points, segments
+
+        else:
 
-        return xing_points, segments
+            # Otherwise, subset valid crossings
+            segments = segments[valid, :]
+            switch_interv = switch_interv[valid, :]
+            left_y = left_y[valid, :]
+            right_y = right_y[valid, :]
 
+            # Find crossing points.
+            xing_points = [
+                calc_linear_crossing(switch_interv[i, :], left_y[i, :], right_y[i, :])
+                for i in range(segments.shape[0])
+            ]
 
-def calc_prim_kink(mGrid, vTGrids, choices):
+            xing_array = np.asarray(xing_points)
+
+            return xing_array, segments
+
+
+def calc_nondecreasing_segments(x, y):
```

Comment:
I'd like to ask you something re parallel and numba. Do you know how the parallel decorator works internally? Like how many cores does it take by default, is that option set somewhere else? Is it smart about figuring out when it is worth it to do parallel?

The operation that is performed inside the loop is so simple that I would expect the overhead from parallelization being greater than the speed gains. I might be wrong, because I have never parallelized using numba, but I am interested in your experience with it.

Reply:
Parallel or not parallel, I've just numbified it. Thanks!

Reply:
Yes, in my experience setting `parallel=True` and using `prange` are smart in that numba has internal heuristics on how to parallelize and how many cores to use. Beyond that, the internal operations were always too obscure for me. 

This, or other uses for numba in general, might be a good use case for @sbenthall's idea of a HARK model profiler #982.